test/integration/endpointslice/endpointslicemirroring_test.go

```diff
@@ -228,3 +231,213 @@ func TestEndpointSliceMirroring(t *testing.T) {
+			_, err = client.CoreV1().Endpoints(ns.Name).Update(context.TODO(), customEndpoints, metav1.UpdateOptions{})
+			if err != nil {
+				t.Fatalf("Error updating endpoints: %v", err)
+			}
+
+			// verify the endpoint updates were mirrored
+			err = wait.PollImmediate(1*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+				lSelector := discovery.LabelServiceName + "=" + service.Name
+				esList, err := client.DiscoveryV1beta1().EndpointSlices(ns.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+				if err != nil {
+					t.Logf("Error listing EndpointSlices: %v", err)
+					return false, err
+				}
+
+				if len(esList.Items) == 0 {
+					t.Logf("Waiting for EndpointSlice to be created")
+					return false, nil
+				}
+				if len(esList.Items) > 1 {
+					return false, fmt.Errorf("Only expected 1 EndpointSlice, got %d", len(esList.Items))
+				}
+				endpointSlice := esList.Items[0]
+				if endpointSlice.ObjectMeta.ResourceVersion == resVersion {
+					t.Logf("Waiting for EndpointSlice to be updated")
+					return false, nil
+				}
+
+				if endpointSlice.Labels[discovery.LabelManagedBy] != "endpointslicemirroring-controller.k8s.io" {
+					return false, fmt.Errorf("Expected EndpointSlice to be managed by endpointslicemirroring-controller.k8s.io, got %s", endpointSlice.Labels[discovery.LabelManagedBy])
+				}
+
+				// compare addresses
+				epAddresses := []string{}
+				for _, address := range customEndpoints.Subsets[0].Addresses {
+					epAddresses = append(epAddresses, address.IP)
+				}
+
+				sliceAddresses := []string{}
+				for _, sliceEndpoint := range endpointSlice.Endpoints {
+					sliceAddresses = append(sliceAddresses, sliceEndpoint.Addresses...)
+				}
+
+				if !assert.ElementsMatch(t, epAddresses, sliceAddresses) {
+					return false, fmt.Errorf("Expected EndpointSlice to have the same IP addresses, expected %v got %v", epAddresses, sliceAddresses)
+				}
+
+				// check labels were mirrored
+				if !isSubset(customEndpoints.Labels, endpointSlice.Labels) {
+					return false, fmt.Errorf("Expected EndpointSlice to mirror labels, expected %v to be in received %v", customEndpoints.Labels, endpointSlice.Labels)
+				}
+
+				// check annotations but endpoints.kubernetes.io/last-change-trigger-time were mirrored
+				annotations := map[string]string{}
+				for k, v := range customEndpoints.Annotations {
+					if k == corev1.EndpointsLastChangeTriggerTime {
+						continue
+					}
+					annotations[k] = v
+				}
+				if !apiequality.Semantic.DeepEqual(annotations, endpointSlice.Annotations) {
+					return false, fmt.Errorf("Expected EndpointSlice to mirror annotations, expected %v received %v", customEndpoints.Annotations, endpointSlice.Annotations)
+				}
+				return true, nil
+			})
+			if err != nil {
+				t.Fatalf("Timed out waiting for conditions: %v", err)
+			}
+		})
+	}
+}
+
+// isSubset check if all the elements in a exist in b
+func isSubset(a, b map[string]string) bool {
```

Comment:
Somehow I'd thought we already had a helper func for this but I can't find it. We do have an [identically named function in admission](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/podnodeselector/admission.go#L263) though...

Reply:
yeah, there is no good library to deal with annotations and labels, labels have some utils, maybe is a good first issue to create this in apimachinery tools